lineapy/graph_reader/artifact_collection.py

```diff
@@ -0,0 +1,335 @@
+                art._session_id, []
+            ) + [art]
+
+        # For each session, construct SessionArtifacts object
+        for session_id, session_artifacts in artifacts_by_session.items():
+            self.session_artifacts[session_id] = SessionArtifacts(
+                session_artifacts
+            )
+
+    def _sort_session_artifacts(
+        self, dependencies: TaskGraphEdge = {}
+    ) -> List[SessionArtifacts]:
+        """
+        Use the user-provided artifact dependencies to
+        topologically sort a list of SessionArtifacts objects.
+        Raise an exception if the graph contains a cycle.
+        """
+        # Construct a combined graph across multiple sessions
+        combined_graph = nx.DiGraph()
+        for session_artifacts in self.session_artifacts.values():
+            session_graph = session_artifacts.graph.nx_graph
+            combined_graph.add_nodes_from(session_graph.nodes)
+            combined_graph.add_edges_from(session_graph.edges)
+
+        # Augment the graph with user-specified edges
+        dependency_edges = list(
+            chain.from_iterable(
+                (
+                    (
+                        self.art_name_to_node_id.get(artname, None),
+                        self.art_name_to_node_id.get(to_artname, None),
+                    )
+                    for artname in from_artname
+                )
+                for to_artname, from_artname in dependencies.items()
+            )
+        )
+        if None in list(chain.from_iterable(dependency_edges)):
+            raise KeyError(
+                "Dependency graph includes artifacts not in this artifact collection."
+            )
+        combined_graph.add_edges_from(dependency_edges)
+
+        # Check if the graph is acyclic
+        if nx.is_directed_acyclic_graph(combined_graph) is False:
+            raise Exception("Provided dependencies result in a cyclic graph.")
+
+        # Identify topological ordering between sessions
+        session_id_nodes = list(self.session_artifacts.keys())
+        session_id_edges = []
+        for node_id, to_node_id in dependency_edges:
+            assert node_id is not None
+            assert to_node_id is not None
+            from_session_id = self.node_id_to_session_id.get(node_id, None)
+            to_session_id = self.node_id_to_session_id.get(to_node_id, None)
+            if from_session_id is not None and to_session_id is not None:
+                session_id_edges.append((from_session_id, to_session_id))
+        inter_session_graph = nx.DiGraph()
+        inter_session_graph.add_nodes_from(session_id_nodes)
+        inter_session_graph.add_edges_from(session_id_edges)
+        session_id_sorted = nx.topological_sort(inter_session_graph)
+
+        return [
+            self.session_artifacts[session_id]
+            for session_id in session_id_sorted
+        ]
+
+    @staticmethod
+    def _extract_session_module(
```

Comment:
@yoonspark  real quick can you just confirm which of the following functions are moved from session_artifacts.py 

why they were moved, and which ones (if any) need any additional review and which ones are unchanged

Reply:
@andycui97 `_extract_session_module` is the method that was moved/adapted from `SessionArtifacts`. This decision was made because `ArtifactCollection` is the level where we want code modularization to happen (as the user may want modularization with artifacts across multiple sessions). So, rather than scattering code modularization/generation in both `SessionArtifacts` and `ArtifactCollection`, we decided to do this work in the latter.

Other than this static method, other methods are rather unique to `ArtifactCollection`, so they need review.